### PR TITLE
[lldb] Introduce precise Fallback SwiftASTContexts. 

### PR DESCRIPTION
**Pull request by adrian-prantl**
When we switched to precise compiler invocations, we still kept the per-module fallback SwiftASTContext. This can be a problem when also using explicit modules, because explicit module imports only work with precise compiler invocations. This patch changes
TypeSystemSwiftTyperef to pass its SymbolContext when creating its fallback SwiftASTContext. This basically turns on precise compiler invocationsfor fallback contexts. The price is more compiler instances, but this is only way to make them work reliably with explicit modules.

rdar://137087616

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 1]
```diff
@@ -203,8 +203,11 @@ SwiftUnsafeRawBufferPointer::SwiftUnsafeRawBufferPointer(ValueObject &valobj)
 }
 
 lldb::ChildCacheState SwiftUnsafeRawBufferPointer::Update() {
-  if (!m_valobj.GetNumChildren())
+  auto success = m_valobj.GetNumChildren();
```
Comment (augusto2112) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: Could you spell out the type of `success` or change the name to something like `num_children_or_error`? It's not obvious in the `if` statement below that you're checking for the error, and not that the number of children is not 0.

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 2]
```diff
@@ -2659,58 +2670,94 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
...
-  Target &target = *target_sp;
 
   // Make an AST but don't set the triple yet. We need to
   // try and detect if we have a iOS simulator.
-  std::shared_ptr<SwiftASTContextForExpressions> swift_ast_sp(
-      new SwiftASTContextForExpressions(m_description, typeref_typesystem));
+  std::shared_ptr<SwiftASTContext> swift_ast_sp;
+  if (for_expressions) {
+    // Expression context.
+    if (!target_sp && for_expressions) {
```
Comment (augusto2112) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp]: If we're here that means `for_expressions` is true

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 3]
```diff
@@ -2659,58 +2670,94 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
...
+      LOG_PRINTF(GetLog(LLDBLog::Types), "No target for expression typesystem");
+      return {};
+    }
+    swift_ast_sp.reset(new SwiftASTContextForExpressions(
+        m_description,
+        *llvm::cast<TypeSystemSwiftTypeRef>(&typeref_typesystem)));
+    // This is a scratch AST context, mark it as such.
+    swift_ast_sp->m_is_scratch_context = true;
+
+    swift_ast_sp->GetLanguageOptions().EnableCXXInterop =
```
Comment (augusto2112) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp]: Looks like there's some repeated code here, could you extract that to a `SetCompilerOptions` function?

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 4]
```diff
@@ -2659,58 +2670,94 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
...
-  swift_ast_sp->GetLanguageOptions().EnableCXXInterop =
-      ShouldEnableCXXInterop(cu);
-
-  if (target.IsEmbeddedSwift())
-    swift_ast_sp->GetLanguageOptions().enableFeature(swift::Feature::Embedded);
-
   bool handled_sdk_path = false;
-  const size_t num_images = target.GetImages().GetSize();
+  ModuleList module_module;
+  if (!target_sp)
```
Comment (augusto2112) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:2743]: We have a lot of code checking if there's no target in this function which didn't exist before. Is this even possible? Could we be instantiating an AST context without a target?

Should we just return early if we don't have a target?

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 5]
```diff
@@ -2633,12 +2634,20 @@ static lldb::ModuleSP GetUnitTestModule(lldb_private::ModuleList &modules) {
   return ModuleSP();
 }
 
-lldb::TypeSystemSP SwiftASTContext::CreateInstance(
-    const SymbolContext &sc,
-    TypeSystemSwiftTypeRefForExpressions &typeref_typesystem,
-    const char *extra_options) {
+lldb::TypeSystemSP
+SwiftASTContext::CreateInstance(const SymbolContext &sc,
```
Comment (augusto2112) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:2638]: Reading the code, this function is huge and does a lot of different things. We should probably split it up at some point, but not necessarily in this patch.

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp (hunk 3 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp]: I cleaned this up a bit, factoring anything out further into function makes the code less readable.

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp (hunk 4 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:2743]: This function is called for 
- creating an SwiftASTContextForExpression (with target)
- creating a precise per-module fallback SwiftASTContext for a TypeSystemSwiftTypeRef (no target)

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp (hunk 5 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:2638]: Yes, functions like the triple computation could get factored out for readability.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Pull request merged**